aptos-move/aptos-vm-types/src/storage/space_pricing.rs

```diff
@@ -127,8 +150,187 @@ impl DiskSpacePricing {
             },
             Deletion => ChargeAndRefund {
                 charge: 0.into(),
-                refund: metadata.total_deposit().into(),
+                refund: op.metadata_mut.total_deposit().into(),
+            },
+        }
+    }
+
+    fn charge_refund_write_op_v2(
+        params: &TransactionGasParameters,
+        op: WriteOpInfo,
+    ) -> ChargeAndRefund {
+        use WriteOpSize::*;
+
+        let key_size = op.key.size() as u64;
+        let num_bytes = key_size + op.op_size.write_len().unwrap_or(0);
+        let target_bytes_deposit: u64 = num_bytes * u64::from(params.storage_fee_per_state_byte);
+
+        match op.op_size {
+            Creation { .. } => {
+                // permanent storage fee
+                let slot_deposit = u64::from(params.storage_fee_per_state_slot);
+
+                op.metadata_mut.maybe_upgrade();
+                op.metadata_mut.set_slot_deposit(slot_deposit);
+                op.metadata_mut.set_bytes_deposit(target_bytes_deposit);
+
+                ChargeAndRefund {
+                    charge: (slot_deposit + target_bytes_deposit).into(),
+                    refund: 0.into(),
+                }
+            },
+            Modification { write_len } => {
+                // Change of slot size or per byte price can result in a charge or refund of the bytes fee.
+                let old_bytes_deposit = op.metadata_mut.bytes_deposit();
+                let state_bytes_charge = if write_len > op.prev_size as u64
+                    && target_bytes_deposit > old_bytes_deposit
+                {
+                    let charge_by_increase: u64 = (write_len - op.prev_size as u64)
+                        * u64::from(params.storage_fee_per_state_byte);
+                    let gap_from_target = target_bytes_deposit - old_bytes_deposit;
+                    std::cmp::min(charge_by_increase, gap_from_target)
+                } else {
+                    0
+                };
+                op.metadata_mut.maybe_upgrade();
+                op.metadata_mut
+                    .set_bytes_deposit(old_bytes_deposit + state_bytes_charge);
+
+                ChargeAndRefund {
+                    charge: state_bytes_charge.into(),
+                    refund: 0.into(),
+                }
+            },
+            Deletion => ChargeAndRefund {
+                charge: 0.into(),
+                refund: op.metadata_mut.total_deposit().into(),
             },
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use aptos_types::{
+        on_chain_config::CurrentTimeMicroseconds, state_store::state_value::StateValueMetadata,
+    };
+
+    #[test]
+    fn test_bytes_deposit() {
+        let pricing = DiskSpacePricing::V2;
+        let mut params = TransactionGasParameters::random();
```

Comment:
Is there a reason why these parameters are set randomly? What I'm worried about is that randomly generated values will rarely make any sense for certain gas parameters, and may even cause the VM to have weird behaviors (e.g. rejecting all transactions). It's also generally not a good idea to have non-determinism in a test.

Reply:
To test that the charging logic only relies on the dimensions I set manually here (to be zero or non-zero).
The test doesn't use the VM so it was not a concern for me. 
Do you think it's a better idea to have `params.with_all(32432)`?

Reply:
Would you be fine with `TransactionGasParamters::zeros()`?

Reply:
That's exactly what I want to avoid - something that the logic accidentally relies on, but because it's zero, the tests don't show its effect.

Reply:
No what I meant was that maybe you could start with `TransactionGasParamters::zeros()`, and then set the few parameters you care about to non-zero values. Will it work?

Reply:
if `DiskSpacePricing::charge_event()` accidentally charged `params.fee_per_txn_byte * event.size()` and I didn't know or set it to non-zero in the test, I wouldn't detect it. Does it make sense?